bots/tests-data

```diff
@@ -0,0 +1,254 @@
+    opts = parser.parse_args()
+
+    for pull in pulls(opts.open and "open" or "closed", since=opts.since):
+        if opts.open:
+            merged = None
+        else:
+            merged = True
+        if opts.verbose:
+            sys.stderr.write("pull-{0}\n".format(pull["number"]))
+        for revision in revisions(pull):
+            if opts.verbose:
+                sys.stderr.write("- {0}\n".format(revision))
+            for (context, created, url, log) in logs(revision):
+                if opts.verbose:
+                    sys.stderr.write("  - {0} {1}\n".format(created, context))
+                for (status, name, body) in tap(log):
+                    sys.stdout.write(json.dumps({
+                        "pull": pull["number"],
+                        "revision": revision,
+                        "status": status,
+                        "context": context,
+                        "date": created,
+                        "head": merged,
+                        "test": name,
+                        "url": url,
+                        "log": body
+                    }))
+                    sys.stdout.write("\n")
+            if merged:
+                merged = False
+
+
+# An HTML parser that just pulls out all the <a href="...">
+# link hrefs in a given page of content. We also qualify these
+# hrefs with a base url, in case they're relative
+class HrefParser(HTMLParser.HTMLParser):
+    def __init__(self, base, hrefs):
+        HTMLParser.HTMLParser.__init__(self)
+        self.hrefs = hrefs
+        self.base = base
+
+    def handle_starttag(self, tag, attrs):
+        if tag.lower() == "a":
+            for (name, value) in attrs:
+                if name.lower() == "href":
+                    url = urlparse.urljoin(self.base, value)
+                    # print 'HREF', url
+                    self.hrefs.append(url)
+
+# Retrieves the content of the given URL
+def retrieve(url):
+    return urllib2.urlopen(url).read()
+
+# Returns a list of all results at the given URL
+def links(url):
+    result = [ ]
+    parser = HrefParser(url, result)
+    try:
+        parser.feed(retrieve(url))
+    except urllib2.HTTPError, ex:
+        if ex.code != 404:
+            raise
+    except urllib2.URLError:
+        pass
+    return result
+
+# Generates full GitHub JSON closed pull request
+# objects since the given date. We only care about closed
+# pull requests for processing, the rest are still in flux.
+def pulls(state='closed', since=None):
+    for pull in task.api.pulls(state=state, since=since):
+        yield pull
+
+# Generates revisions for a given pull request. Each revision
+# is a simple string sha. The first one is the one that is at
+# the head of the pull request, and any other follow on revisions
+# are those which were previously tested (eg: force push, or with
+# later fixups).
+def revisions(pull):
+    head = pull["head"]["sha"]
+    if not head:
+        return
```

Comment:
Return an empty list here or `raise StopIteration`, as the caller unconditionally iterates over the result. However, is that actually an expected state? A PR should *always* have a HEAD and a SHA for it, no? Maybe turn this into an assertion instead? This would already `KeyError` if the PR doesn't have a `["head"]`.

Reply:
Fixed for the theoretical KeyError.

I think that just using return is more pythonic. See: https://stackoverflow.com/questions/14183803/in-pythons-generators-what-is-the-difference-between-raise-stopiteration-and

Reply:
Thanks, I wasn't aware that a bare return works for generators, looks much simpler indeed!

I didn't actually mean that the `KeyError` should be fixed - it's good to have it as a kind of internal assertion. I just wondered if it can actually happen that a HEAD has no SHA.